File: 1888. minimum-number-of-flips-to-make-the-binary-string-alternating.py
class Solution:
    def minFlips(self, s):
        n = len(s)
        e, o = (n + 1) // 2, n // 2
        x = s[::2].count('1') - s[1::2].count('1')
        r = min(e - x, o + x)
        if n & 1:
            for c in s:
                x = 2 * (ord(c) & 1) - x
                r = min(r, e - x, o + x)
        return r

if __name__ == "__main__":
    sol = Solution()
    
    # Example 1:
    s = "111000"
    output = sol.minFlips(s)
    print(f"Example 1: Input: s = {s}")
    print(f"Output: {output}")
    print(f"Expected: 2")
    print()
    
    # Example 2:
    s = "010"
    output = sol.minFlips(s)
    print(f"Example 2: Input: s = {s}")
    print(f"Output: {output}")
    print(f"Expected: 0")
    print()
    
    # Example 3:
    s = "1110"
    output = sol.minFlips(s)
    print(f"Example 3: Input: s = {s}")
    print(f"Output: {output}")
    print(f"Expected: 1")

File: 1888. minimum-number-of-flips-to-make-the-binary-string-alternating.java
class Solution {
    public static int minFlips(String s) {
        final int n = s.length();
        int p = 0, k = 0;
        for (int i = 0; i < n; i++) {
            final char c = s.charAt(i);
            k += p ^ c & 1;
            p ^= 1;
        }
        int r = Math.min(k, n - k);
        if ((n & 1) != 0) {
            p = 0;
            for (int i = 0; i + 1 < n; i++) {
                final char c = s.charAt(i);
                k += (p ^ c & 1 ^ 1) - (p ^ c & 1);
                p ^= 1;
                r = Math.min(r, Math.min(k, n - k));
            }
        }
        return r;
    }
    
    public static void main(String[] args) {
        Solution sol = new Solution();
        
        // Example 1:
        String s1 = "111000";
        int output1 = minFlips(s1);
        System.out.println("Example 1: Input: s = " + s1);
        System.out.println("Output: " + output1);
        System.out.println("Expected: 2");
        System.out.println();
        
        // Example 2:
        String s2 = "010";
        int output2 = minFlips(s2);
        System.out.println("Example 2: Input: s = " + s2);
        System.out.println("Output: " + output2);
        System.out.println("Expected: 0");
        System.out.println();
        
        // Example 3:
        String s3 = "1110";
        int output3 = minFlips(s3);
        System.out.println("Example 3: Input: s = " + s3);
        System.out.println("Output: " + output3);
        System.out.println("Expected: 1");
    }
}